### ISSUE TITLE
Table clipping is broken with VS2013 (ub or compiler issue?)

### ISSUE THREAD

**Issue by acmepjz** [ocornut/imgui#6377]
**Version/Branch of Dear ImGui:**

```
Dear ImGui 1.89.6 WIP (18954)
--------------------------------
sizeof(size_t): 4, sizeof(ImDrawIdx): 2, sizeof(ImDrawVert): 20
define: __cplusplus=199711
define: _WIN32
define: _MSC_VER=1800
define: IMGUI_HAS_VIEWPORT
define: IMGUI_HAS_DOCK
--------------------------------
io.BackendPlatformName: imgui_impl_sdl2
io.BackendRendererName: imgui_impl_opengl3
io.ConfigFlags: 0x00000043
 NavEnableKeyboard
 NavEnableGamepad
 DockingEnable
io.ConfigViewportsNoDecoration
io.ConfigInputTextCursorBlink
io.ConfigWindowsResizeFromEdges
io.ConfigMemoryCompactTimer = 60.0
io.BackendFlags: 0x00001C06
 HasMouseCursors
 HasSetMousePos
 PlatformHasViewports
 HasMouseHoveredViewport
 RendererHasViewports
--------------------------------
io.Fonts: 1 fonts, Flags: 0x00000000, TexSize: 512,64
io.DisplaySize: 1600.00,837.00
io.DisplayFramebufferScale: 1.00,1.00
--------------------------------
style.WindowPadding: 8.00,8.00
style.WindowBorderSize: 1.00
style.FramePadding: 4.00,3.00
style.FrameRounding: 0.00
style.FrameBorderSize: 0.00
style.ItemSpacing: 8.00,4.00
style.ItemInnerSpacing: 4.00,4.00
```

**My Issue/Question:**

The vertical scrolling with clipping in the demo window seems not working correctly. The lower clipping seems correct, but the upper clipping is wrong. Tested on 1.89.5 WIP and 1.89.6 WIP of docking branch, both have problems. Only tested sdl2+opengl3 backend on Windows. Normal branch not tested.

Also tested online demo of imgui_bundle (1.89.3) which does not have this problem.

**Screenshots/Video**

![screenshot1](https://user-images.githubusercontent.com/3397779/234729489-fa419ade-d1fe-4900-9b5c-fdc658b82f68.png)

**Comment by ocornut** [comment]
I cannot seem to repro this if I compile `examples/example_sdl2_opengl3/main.cpp`.
Can you confirm you are running the same unmodified example app?

**Comment by acmepjz** [comment]
I tried the unmodified example app, and the problem still exists. I tried sdl2+opengl3 and win32+opengl3 and win32+dx11, all of them have this problem. Besides the docking branch, I also tried the default branch, and the problem still exists.

One thing I have to mention is that I'm using VS2013 which does not support `constexpr`. So I have to add `/Dconstexpr=` to the command line switches. I don't know if it will introduce bugs.
 
![002](https://user-images.githubusercontent.com/3397779/234888336-e7632520-36e8-4d73-ac83-743dd043cf5f.png)

![003](https://user-images.githubusercontent.com/3397779/234888340-e9e6ca0d-3abc-4922-afb8-6e21824bbf7d.png)

![004](https://user-images.githubusercontent.com/3397779/234888344-42ed2c39-76b1-436f-ba9c-2c2dcf0085c8.png)

**Comment by ocornut** [comment]
Can you got under Metrics->DrawList and look at the clipping rectangle for those draw commands emitting the text inside the table?

**Comment by acmepjz** [comment]
Sure.

![111](https://user-images.githubusercontent.com/3397779/234941171-2d595375-a6aa-4f01-bb27-45090176212a.png)

As a comparison, this is from the official demo binary:

![222](https://user-images.githubusercontent.com/3397779/234941218-805b0a5f-7690-42cb-a1fe-e4c7ddb6e339.png)

I think maybe it's compiler specific. I'll try to find out it later.

**Comment by acmepjz** [comment]
That's very strange, the same code compiled with VS2015 does not have this problem. 😅 Also I think the problem is in `TableMergeDrawChannels()` since if I change this function to return immediately then the problem is gone. I'll do further investigation.

**Comment by ocornut** [comment]
Interesting. Indeed since we dropped 2012 I started testing with VS 2015 as a minimum target and not VS 2013.
That code in `TableMergeDrawChannels()` is rather low-level and may be accidentally relying on undefined behavior. In this case it is accidentally merging draw calls where it shouldn't.

There's debug code in the middle of the function, try to change the `#if 0` into `#if 1` to investigate.

**Comment by ocornut** [comment]
I installed VS2013 (that my 8th VS toolset variant on the machine) and confirmed the bug.

This struct on stack was not properly initialized:
```cpp
    struct MergeGroup
    {
        ImRect          ClipRect;
        int             ChannelsCount;
        ImBitArrayPtr   ChannelsMask;
    };
    int merge_group_mask = 0x00;
    MergeGroup merge_groups[4] = {};
```
I don't know what happened during those years when VS was working on their C++11 conformance, and I may not know the standard enough: I assumed `={};` would zero-init anything that doesn't have an explicit constructor but specifically VS2013 doesn't seem to do it. Presumably the mixture of types with constructors and types without are throwing this off.

Pushed a fix 7c291ba.

Search and didn't find other occurrences of this pattern in the code.

I am not adding a `#define constexpr` in the codebase as a signal that this version is in theory not supported, as I don't think VS13 was fully C++11 conformant, but I think it'll work and we are not likely to use a lots more C++ features in the future.
